File: Apps_C/P04-MultipleSensors/APDS9300.c
#include "APDS9300.h"
#include "i2c.h"
#include "math.h"

#pragma GCC diagnostic ignored "-pedantic"

/// \defgroup light Ambient Light Sensor 
/// These functions let you communicate with the APDS9300 ambient light sensor
/// @{

/**
 *@brief Powers on the sensor , sets sensor gain to 1 and the sampling time to 402 ms and disables the interrupts
 *@param none
 *@return devId The device ID for the abient light sensor
 */
unsigned char AL_Initialize(void)
{
    AL_PowerState(POWER_ON);				//Power on sensor	
	char devId = APDS9300_ReadByte(COMMAND | CONTROL);			//Value read should be 0x13
	//AL_Clear_Interrupt();					//Clear interrupts
    //AL_SetGain(GAIN_1);                 	//Sets gain to 1
    //AL_SetSamplingTime(INTEG402MS);		//Set sampling time to 402 ms
	
	return devId;
}

/**
 *@brief Set the sensor power state 
 *@param state	variable fo type powerState denotes the power state of the sensor
 *@return none
 */
void AL_PowerState(powerState state)
{
	//unsigned char cmd  = state ? POWER_ON : POWER_OFF;
    APDS9300_WriteByte(0x60);			//Write 0x60 to clear and configure for word transactions	
	APDS9300_WriteByte(0x03);			//Write 0x03 to CONTROL register to power on the sensors
}

/**
 *@brief Returns the CHIP id.
 *@param none
 *@return partID Chip ID
 */
unsigned char AL_ChipID(void)
{
    unsigned char partID = 0x00;

    partID = APDS9300_ReadByte(COMMAND|ID);
    return partID;
}

/**
 *@brief Reads the value from one of the two channels
 *@param channel Channel id
 *@return channelValue 16 bit sensor data representing channel value
 */
unsigned int AL_ReadChannel(channel chan)
{
    unsigned int channelValue = 0;
    switch(chan)
    {
        case CH0:
				APDS9300_WriteByte(COMMAND|CMD_CLEAR_INT|CMD_WORD|DATA0LOW);
                channelValue = APDS9300_ReadWord();
                break;
        case CH1:
				APDS9300_WriteByte(COMMAND|CMD_CLEAR_INT|CMD_WORD|DATA1LOW);
                channelValue = APDS9300_ReadWord();
                break;
        default:

        break;
    }

    return channelValue;
}

/**
 * @brief Returns LUX output from the sensor
 * @param ch1 Channel 1 value
 * @param ch2 Channel 2 value
 * @return Lux Float value of light intensity
 */
float AL_Lux(unsigned int ch0, unsigned int ch1)
{
	float k = ch1/ch0;
	float Lux=0;

	if((k>=0)&& (k<=0.52))
		Lux=(0.0315*ch0)-(0.0593*ch0*pow(k,1.4));
	else if((k>0.52)&& (k<=0.65))
		Lux=(0.0229*ch0)-(0.0291*ch1);
	else if((k>0.65)&& (k<=0.80))
		Lux=(0.0157*ch0)-(0.0180*ch1);
	else if((k>0.80)&& (k<=1.30))
		Lux=(0.00338*ch0)-(0.00260*ch1);
	else 
		Lux=0;
	return Lux;
}

/**
 *@brief Set the sensor gain. Default gain value is GAIN_1.
 *@param val Gain type can be GAIN_1 or GAIN_16 for gain of 1 or 16x
 *@return val Gain mode
 */
unsigned char AL_SetGain(gain val)
{
    unsigned char value = 0;
    switch(val)
    {
        case GAIN_1:
            APDS9300_WriteByte(COMMAND|TIMING);       //Write to TIMING register
			APDS9300_WriteByte(GAIN);
            break;
        case GAIN_16:
            value  = APDS9300_ReadByte(TIMING);       //Write to TIMING register
            value &= ~GAIN;
            APDS9300_WriteByte(TIMING|COMMAND);
			APDS9300_WriteByte(value);
            break;

        default:
            APDS9300_WriteByte(COMMAND|TIMING);
			APDS9300_WriteByte(GAIN);
    }
    return val;
}

/**
 *@brief Sets the sampling time for the sensor,can be one of three predetermined values
 *@param sampling_time can be 0x00,0x01,0x02
 *@return none
 */
void AL_SetSamplingTime(unsigned char sampling_time)
{
    APDS9300_WriteByte(TIMING|COMMAND);
	APDS9300_WriteByte(sampling_time);
}

/**
 *@brief Sets the low threshold value for the interrupt
 *@param lowthreshvalue Interrupt low threshold value
 *@return none
 */
void AL_SetIntLowThreshold(unsigned int lowthreshvalue)
{
     APDS9300_WriteWord(THRESHLOWLOW| COMMAND | CMD_WORD ,lowthreshvalue);	//use the write word protocol to write these 16 bit values
}

/**
 *@brief Sets the high threshold value for the interrupt
 *@param highthreshvalue Interrupt high threshvalue value
 *@return none
 */
void AL_SetIntHighThreshold(unsigned int highthreshvalue)
{
    APDS9300_WriteWord(THRESHHIGHLOW| COMMAND | CMD_WORD ,highthreshvalue);
}

/**
 *@brief Clear any existing interrupts
 *@param none
 *@return none
 */
void AL_Clear_Interrupt(void)
{
    APDS9300_WriteByte(COMMAND|CMD_CLEAR_INT);			//Set clear bit to 1
}

/**
 *@brief Enables or disables the interrupt and sets the persistence of interrupt occurences
 *		 The interrupt will fire after x occurences outside the threshold values
 *		 Default values are interrupts enabled and interrupts enabled
 *@param enable 	  Enables or disable interrupts 1/0 to enable/disable
 *@param persistence  Value from 0 to 16 for value 0 interrupts occur after every sample
 *@return none
 */
void AL_ConfigureInterrupt(unsigned char enable, unsigned char persistence)
{
    if(enable)
    {
        //APDS9300_WriteByte((COMMAND|INTERRUPT|INTERR_ENA) & persistence);
    }
    else
    {
        APDS9300_WriteByte(COMMAND|INTERRUPT|INTERR_DIS);
    }
}

/// @}


/**
 * @brief Writes a single byte of data to the current sensor register. 
 * 		  Function assumes that I2C peripheral is preconfigured.
 * @param data Data byte to write to the sensor.
 * @return none
 */
void APDS9300_WriteByte(unsigned char data)
{
	I2C_WriteByte(data);			//Sets a specific register to a certain value
}

/**
 * @brief Writes a byte to the specific register
 * @param reg Address of sensor register 
 * @param data Value to be written on the sensor register.
 * @return none
 */
void APDS9300_WriteRegister(unsigned char reg, unsigned char data)
{
	I2C_WriteByteRegister(reg,data);			//Writes data on the specific register
}

/**
 * @brief Writes a word to the specific register
 * @param reg  Address of sensor register
 * @param data 16 bit data  to be written on the sensor register. 
 * @return none
 */
void APDS9300_WriteWord(unsigned char reg, unsigned int data)
{
	unsigned char buffer[2];
	buffer[1] = (unsigned char) (data >> 8);			//MSB
	buffer[0] = (unsigned char) (data && 0xff);			//LSB
	I2C_WriteWordRegister(reg,buffer);
}

/**
 * @brief Reads a single byte from the given sensor register
 * @param reg Address of sensor register
 * @return Val Value of sensor register
 */
char APDS9300_ReadByte(char reg)
{
    return I2C_ReadByteRegister(reg);
}

/**
 * @brief Read word from register
 * @param reg Register address
 * @return val 16bit word content of register
 */
unsigned int APDS9300_ReadWordReg(char reg)
{	unsigned int val = I2C_ReadWordRegisterRS(reg);
    return val;
}

/**
 * @brief Read a single word from the sensor. Assumes register pointer is preset.
 * @param none
 * @return value 16bit word read from the preset register
 */
unsigned int APDS9300_ReadWord(void)
{
	char buff[2] = {0x00,0x00};
	bcm2835_i2c_read(buff,2);

	return (buff[1] << 8)|buff[0];
}

File: Drivers_Python/APDS-9300/APDS9300.py
#!/usr/bin/python
import sys
import smbus
import time
bus = smbus.SMBus(1)

"""
__author__ = "D.Qendri"
__copyright__ = "Copyright 2015 Sensorian"
__license__ = "GPL V3"
__version__ = "1.0"
"""

APDS9300ADDR = 0x29					#Address of ambient light sensor

#Register definitions
CONTROL           = 0x00			#Control of basic functions
TIMING            = 0x01
THRESHLOWLOW      = 0x02
THRESHLOWHIGH     = 0x03
THRESHHIGHLOW     = 0x04
THRESHHIGHHIGH    = 0x05
INTERRUPT         = 0x06			#Interrupt control 
CRC               = 0x08
ID                = 0x0A
DATA0LOW          = 0x0C
DATA0HIGH         = 0x0D
DATA1LOW          = 0x0E
DATA1HIGH         = 0x0F

#COMMAND REG
COMMAND           = 0x80
CMD_CLEAR_INT     = 0x40
CMD_WORD          = 0x20

#CONTROL REG
POWERON 		  = 0x03
POWEROFF          = 0x00

#Timing Reg
GAIN              = 0x10
MANUAL            = 0x08
INTEG13_7MS       = 0x00
INTEG101MS        = 0x01
INTEG402MS        = 0x02

#INTERRUPT REG
INTERR_THRSHOLD   = 0x01
INTERR_DIS        = 0x00
INTERR_EN         = 0x10

class APDS9300(object):
	""" 
	Ambient Light Sensor. Class functions for the APDS9300 ambient light sensor. 
	"""
	
	def __init__(self):
		"""
		Powers on the sensor 
		
		
		:param none: 
		:returns: none
		"""
		self._address = APDS9300ADDR
		self.setpowerState(POWERON)				#Power on sensor
		self.setGain(1)
		self.setSamplingTime(0x02)

	def setpowerState(self,state):
		"""
		Configures the power state of the chip.
		
		
		:param state:  Denotes the power state of the sensor ,set to 0/1 to turn off/on.
		:returns: none
		"""
		self.writeRegister(CONTROL|CMD_CLEAR_INT|CONTROL,state)		#Turn on or off the sensor

	def setGain(self,gain):
		"""
		Set the gain of the internal photodiode sensors, can be 1x or 16x.
		
		
		:param gain: Gain type can be GAIN_1 or GAIN_16 for gain of 1 or 16x
		:returns: none
		"""
		value = 0
		if (gain == 1):
			value  = self.readByte(COMMAND|TIMING)      #Write to TIMING register
			value &= 0b11101111
			self.writeByte(COMMAND|TIMING)       		#Write to TIMING register
			self.writeByte(value)
		elif (gain == 16):
			self.writeByte(COMMAND|TIMING)				#Write to TIMING register
			self.writeByte(GAIN)
		else :
			self.writeByte(COMMAND|TIMING)				#Write to TIMING register
			self.writeByte(GAIN)

	def setSamplingTime(self,sampling_time):
		"""
		Configure the sampling time of the sensor.
		
		
		:param sampling_time: Can be 0x00,0x01,0x02
		:returns: none
		"""
		self.writeByte(TIMING|COMMAND)
		self.writeByte(sampling_time)

	def readChannel(self,channel):
		"""
		Read raw data from channel 0 or channel 1
		
		
		:param channel: Value of 0 or 1.
		:returns data : Channel data as raw number.
		"""
		if(channel == 1):
			self.writeByte(COMMAND|CMD_CLEAR_INT|CMD_WORD|DATA0LOW)
			channel = self.readWord(0x00)
		else :
			self.writeByte(COMMAND|CMD_CLEAR_INT|CMD_WORD|DATA1LOW)
			channel = self.readWord(0x00)
		return channel

	def getLuxLevel(self,ch0, ch1):
		"""
		Returns LUX output from the raw sensor data.
		
		
		:param ch0: Channel 0 value 
		:param ch1: Channel 1 value 
		:returns: Lux Float value of light intensity
		"""
		k = 0
		if(ch0 !=0):
			k = float(ch1/ch0)
			
		Lux=0

		if((k>=0) and (k<=0.52)):
			Lux=(0.0315*ch0)-(0.0593*ch0*pow(k,1.4))
		elif( (k>0.52) and (k<=0.65)):
			Lux=(0.0229*ch0)-(0.0291*ch1)
		elif((k>0.65) and (k<=0.80)): 
			Lux=(0.0157*ch0)-(0.0180*ch1)
		elif((k>0.80) and (k<=1.30)):
			Lux=(0.00338*ch0)-(0.00260*ch1)
		else :
			Lux=0
		return Lux
		
	def setIntLowThreshold(self,lowvalue):
		"""
		Set the low level data threshold for the interrupt.
		
		
		:param lowvalue: Low interrupt threshold value 
		:returns: none
		"""
		self.writeWord((THRESHLOWLOW|COMMAND|CMD_WORD),lowvalue)	#use the write word protocol to write these 16 bit values

	def setIntHighThreshold(self,highvalue):
		"""
		Set the high level data threshold for the interrupt.
		
		
		:param highvalue: High interrupt threshold value 
		:returns: none
		"""
		self.writeWord((THRESHHIGHLOW| COMMAND | CMD_WORD),highvalue)

	def clearInterrupt(self):
		"""
		Clear any active interrupts.
		
		
		:param: none
		:returns: none 
		"""
		self.writeByte(COMMAND|CMD_CLEAR_INT)			#Set clear bit to 1

	def setInterruptState(self,enable=False):
		"""
		Enable or disable the interrupts on the sensor.
		
		
		:param enable: Enables or diable interrupt. Default is False.
		:returns: none 
		"""
		if enable == True:
			self.writeRegister(COMMAND|INTERRUPT,INTERR_EN)
		else:
			self.writeRegister(COMMAND|INTERRUPT,INTERR_DIS)

	def chipID(self):
		"""
		Read the unique chip ID.
		
		
		:param none: 
		:returns: partID ID of chip
		"""
		partID = self.readByte(COMMAND|ID)
		return partID
		
	def writeByte(self,data):
		"""
		Write a byte value. Assumes the register pointer is predefined.
		
		
		:param data: Data byte to write on the I2C bus 
		:returns: none
		"""
		bus.write_byte(self._address,data)
		return 0

	def writeRegister(self,reg,data):
		"""
		Write a byte value to a register.
		
		
		:param reg: Register  address
		:param data: Data byte to be written on register. 
		:returns: none
		"""
		bus.write_byte_data(self._address,reg,data)

	def writeWord(self,reg, data):
		"""
		Write a word value to a register.
		
		
		:param reg: Register  address
		:param data: Data word to be written on the register. 
		:returns: none
		"""
		buffer = [(data >> 8) & 0xFF, data & 0xFF]
		bus.write_i2c_block_data(self._address, reg, buffer)

	def readByte(self,reg):
		"""
		Read a byte value from a register.
		
		
		:param reg: Register  address
		:returns: Register byte content
		"""
		x = bus.read_byte_data(self._address, reg)
		return x

	def readWord(self,reg):
		"""
		Read a word value from a register.
		
		
		:param reg: Register  address
		:returns: Register word content
		"""
		x = bus.read_word_data(self._address, reg)
		return x
		